internal/util/borrowonce/storage.go

```diff
@@ -0,0 +1,142 @@
+/*
+borrowonce provides a storage for lending values with a promise to return them. If the value was not returned no one
+else can borrow it. Here is an ELI5:
+
+Suppose you have a box that has a ball inside. You then have people (threads) that want to play with the ball. The first
+person takes the ball. Then someone else looks in the box and there is no ball inside. They wait for the box to have the
+ball again. Then someone else comes in and also wants the ball. But they came later so they will wait until the ball is
+given to the person before and then wait until that person is done playing with the ball.
+*/
+package borrowonce
+
+import (
+	"fmt"
+	"sync"
+)
+
+/*
+Storage lends values in such a way that the first borrower will immediately get the value, and all borrowers after will
+get the value in the same order they requested it, but only after the previous borrower has returned the value.
+
+This is useful if you generally dont operate on the same key in the map, but if you have to do so threads can wait for
+the previous thread to be finished mutating the value and be sure that their operation on the value is in the correct
+order.
+*/
+type Storage[K comparable, V any] struct {
+	storeMu *sync.Mutex         //nolint:structcheck
+	store   map[K]borrowable[V] //nolint:structcheck
+}
+
+// NewStorage creates a new empty storage
+func NewStorage[K comparable, V any]() Storage[K, V] {
+	return Storage[K, V]{
+		storeMu: &sync.Mutex{},
+		store:   make(map[K]borrowable[V]),
+	}
+}
+
+/*
+Future allows you request a position in the borrow queue and Wait() your turn.
+*/
+type Future[V any] struct {
+	vMu sync.Mutex //nolint:structcheck
```

Comment:
`v` is the value, `vMu` is the mutex guarding the value.

Reply:
how about simplifying it to `mu`?

Reply:
I remember someone suggested me to have the mutex be named `varnameMu` and place it above `varname`. Here `v` is the only field in the struct so I guess it can be renamed then.

Reply:
Looking at this diff the mu seems ambigious:
```diff
        } else {
                lockable.queue[0].v = value
-               lockable.queue[0].vMu.Unlock()
+               lockable.queue[0].mu.Unlock()
                lockable.queue = lockable.queue[1:]
        }
```
I think leaving it as vMu would be better because of this.

Reply:
Just when I look at mu its not entirely clear what its for.

Reply:
okay, let's use vMu, this is not a big problem